### PR TITLE
Indirect - Use D-Range lists across multiple runs

### LINKED ISSUE #18858: Indirect Diffraction - OSIRIS dRange
In OSIRISDiffractionReduction there is an option to input a manual dRange - however this is never used and the dRange always defaults to 1.
This should be simple in terms of going through the code and changing where it defaults to 1 to check whether the option for manual drange has been selected. I'm pretty sure the code for manual drange is all in there already.

### PR DESCRIPTION
**Pull request by brandonhewer**
When the default d-range is overridden in the OSIRISDiffractionReduction algorithm, it had not impact on the used d-range, which continued to be the default. The d-range was also supplied as a single number, when it should be supplied as a list of numbers (respective to each supplied run).

The d-range is now correctly used, and is now supplied as a list to OSIRISDiffractionReduction. The interface has been updated to display this change.

This pull request also involves major refactoring of the OSIRISDiffractionReduction algorithm

**To test:**
1. Navigate to `Interfaces > Indirect > Diffraction`.
2. Ensure `OSIRIS` is selected in the Instrument drop-down menu.
3. Ensure `diffonly` is selected in the Reflection drop-down menu.
4. Enter `119977` into the run number field.
5. Ensure the Use Container checkbox is checked and enter `71987` into the field.
6. Supply a calibration file in the cal file field (calibration file included below).
7. Enter `119963` into the vanadium runs field.
8. Ensure the Manual D-Range checkbox is checked and use text field to enter a number.
9. Run the algorithm and check the *_dRange file, ensuring the entered d-range was used.
10. Repeat steps 1-8, but for step 8, try entering:
- Incorrectly formatted list string (correct format is comma separated list of numbers and number ranges written as 'a-b' where a and b are numbers). The expected result should be an error with message "Provided list, 'ENTERED TEXT', was incorrectly formatted."
- Out of bounds D-Range (e.g. 15). The expected result should be "Supplied d-range, ENTERED RANGE, is out of bounds."

Fixes #18858. 

**Release Notes** 

Added change to the release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/OSIRISDiffractionReduction.py [hunk 1]
```diff
@@ -175,23 +218,58 @@ def get_intersection_of_ranges(range_list):
...
+    """
+    Rebins the specified list to the workspace with the smallest
+    x-range in the list.
+
+    :param workspaces: The list of workspaces to rebin to the smallest.
+    :return:           The rebinned list of workspaces.
+    """
+    smallest_ws = min(workspaces, key=lambda ws: mtd[ws].blocksize())
+
+    return [RebinToWorkspace(
```
Comment (louisemccann) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/OSIRISDiffractionReduction.py]: Could you change this to a for loop? Also maybe change Rebin to a child algorith.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/OSIRISDiffractionReduction.py [hunk 2]
```diff
@@ -257,8 +335,7 @@ def PyInit(self):
 
         # Note that dRange numbers are offset to match the numbering in the OSIRIS manual
         # http://www.isis.stfc.ac.uk/instruments/osiris/documents/osiris-user-guide6672.pdf
-        self.declareProperty('DRange', 1, validator=IntBoundedValidator(1, len(TIME_REGIME_TO_DRANGE) + 1),
-                             doc='Drange to use when DetectDRange is disabled')
+        self.declareProperty('DRange', defaultValue="", doc='Dranges to use when DetectDRange is disabled')
```
Comment (louisemccann) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/OSIRISDiffractionReduction.py]: update the doc string to describe what kind of input you need, ie a list of numbers

**Review by louisemccann — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/OSIRISDiffractionReduction.py [hunk 3]
```diff
@@ -126,22 +173,18 @@ def average_ws_list(ws_list):
...
-    numWorkspaces = len(ws_list)
+    num_workspaces = len(ws_list)
 
     # Compute the average and put into "__temp_avg".
-    __temp_avg = mtd[ws_list[0]]
-    for i in range(1, numWorkspaces):
-        __temp_avg += mtd[ws_list[i]]
-
-    __temp_avg /= numWorkspaces
+    temperatures = map(lambda index: mtd[ws_list[index]], range(1, num_workspaces))
```
Comment (louisemccann) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/OSIRISDiffractionReduction.py]: List comprehension

**Comment by louisemccann** [comment]
could you also add the drange used to the sample logs

**Review by NickDraper — CHANGES_REQUESTED** [review]
Code is fine, but it doesn't come out with data in the d-rage specified (I used 2-6).

We need to understand this better.

**Comment by brandonhewer** [comment]
@NickDraper 
These numbers correspond to the Standard Diffraction Settings for drange on OSIRIS as laid out in the manual. Not sure how much this clears up:
https://www.isis.stfc.ac.uk/Pages/osiris-user-guide.pdf

**Comment by NickDraper** [comment]
Can you add that as a tooltip on the input text box

**Comment by brandonhewer** [comment]
@NickDraper Yes, I'll add that now.

**Comment by brandonhewer** [comment]
@NickDraper Note that there are changes to files which aren't specifically related to this PR -> these are ClangFormat changes.

**Review by NickDraper — APPROVED** [review]

**Comment by NickDraper** [comment]
Good to go once the build servers are happy

**Pull request merged**